### ISSUE TITLE
JSON Schema - ditch object group hack

### ISSUE THREAD

**Issue by ewels** [nf-core/tools#687]
At the moment, our JSON schema use `object`s to group parameters together. As this is not organised in the same hierarchical way as the parameters in the pipeline, we have to "flatten" this in the code to make validation work. This is a bit of a nasty hack and I don't love it. x-ref https://github.com/nf-core/tools/pull/593#issuecomment-645033674

A team lead for JSON Schema, @relequestual, got in touch over twitter to offer to help. He suggested an alternative way to be able to structure parameters in interfaces _without_ using `object` properties. This structure should allow native validation of schema whilst maintaining our ability to functionally organise interfaces.

Example schema:

<details>

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "nf-core/gwas pipeline parameters",
  "description": "A pipeline for Genome Wide Association Studies",
  "type": "object",
  "definitions": {
    "inputOutputOptions": {
      "title": "Input/output options",
      "properties": {
        "input": {
          "type": "string",
          "fa_icon": "fas fa-dna",
          "description": "Input FastQ files.",
          "help_text": "A glob pattern for input FastQ files. Should include at least one asterisk (*). For paired-end data, should contain curly brackets with two patterns differentiating the paired reads e.g. `*_R{1,2}.fastq.gz`"
        },
        "single_end": {
          "type": "boolean",
          "description": "Specifies that the input is single-end reads.",
          "fa_icon": "fas fa-align-center",
          "default": false,
          "help_text": "By default, the pipeline expects paired-end data. If you have single-end data, specify this parameter on the command line when you launch the pipeline. It is not possible to run a mixture of single-end and paired-end files in one run."
        },
        "outdir": {
          "type": "string",
          "description": "The output directory where the results will be saved.",
          "default": "./results",
          "fa_icon": "fas fa-folder-open"
        },
        "email": {
          "type": "string",
          "description": "Email address for completion summary.",
          "fa_icon": "fas fa-envelope",
          "help_text": "An email address to send a summary email to when the pipeline is completed.",
          "pattern": "^([a-zA-Z0-9_\\-\\.]+)@([a-zA-Z0-9_\\-\\.]+)\\.([a-zA-Z]{2,5})$"
        }
      },
      "required": [
        "input"
      ],
      "fa_icon": "fas fa-terminal",
      "description": "Define where the pipeline should find input data and save output data."
    },
    "referenceGenomeOptions": {
      "title": "Generic options",
      "properties": {
        "genome": {
          "type": "string",
          "description": "Name of iGenomes reference.",
          "fa_icon": "fas fa-book",
          "help_text": "If using a reference genome configured in the pipeline using iGenomes, use this parameter to give the ID for the reference. This is then used to build the full paths for all required reference genome files e.g. `--genome GRCh38`."
        },
        "fasta": {
          "type": "string",
          "fa_icon": "fas fa-font",
          "description": "Path to FASTA genome file.",
          "help_text": "If you have no genome reference available, the pipeline can build one using a FASTA file. This requires additional time and resources, so it's better to use a pre-build index if possible."
        },
        "igenomes_base": {
          "type": "string",
          "description": "Directory / URL base for iGenomes references.",
          "default": "s3://ngi-igenomes/igenomes/",
          "fa_icon": "fas fa-cloud-download-alt",
          "hidden": true,
          "help_text": ""
        },
        "igenomes_ignore": {
          "type": "boolean",
          "description": "Do not load the iGenomes reference config.",
          "fa_icon": "fas fa-ban",
          "hidden": true,
          "default": false,
          "help_text": "Do not load `igenomes.config` when running the pipeline. You may choose this option if you observe clashes between custom parameters and those supplied in `igenomes.config`."
        }
      },
      "fa_icon": "fas fa-dna",
      "description": "Options for the reference genome indices used to align reads."
    }
  },
  "allOf": [
    {
      "$ref": "#/definitions/inputOutputOptions"
    },
    {
      "$ref": "#/definitions/referenceGenomeOptions"
    }
  ]
}
```

</details>

The key is to use `definitions` to specify sub-schema, and then use the `allOf` keyword to combine these within the main schema for validation. `allOf` takes a list, so that should be the safest way to maintain the correct order of the sections.

Changing the structure of our schema this late in the day is a bit of a pain, as we need to patch quite a lot of code. But I believe that it is worth it even at this stage as we are about to define a standard. If we can do so in the least hacky way possible, that is sure to save us from future pain.

This issue is for changes in the nf-core/tools repo - we will need corresponding changes in the nf-core/nf-co.re website repo, but they can have their own issues.

In tools we need to update the following:

- [x] The template schema
- [x] The `nf-core schema` functionality (likely all subcommands are affected)
- [x] The `nf-core launch` command

Phil

**Comment by ewels** [comment]
Side note: Fun excuse to use this nice new online validator tool: https://jsonschema.dev/s/WvHg3

**Comment by ewels** [comment]
### Explanation of why we needed a nasty hack (and why we should avoid it if possible).

So with the current schema, the basic structure that we have is as follows (simplified psuedo-code):

```yaml
- Input/output options:
    - input
    - single_end
- Reference genome options:
    - genome
    - fasta
```

This is good for building the user interfaces, as it allows us to group the parameters together logically for simpler navigation. The top level groups are defined as `properties` with `type: "object"`.

The problem with this comes when we want to validate the schema though. A valid set of parameters needs to mirror the structure of the schema, so would need to look like this:

```json
{
  "Input/output options": {
    "input": "input.csv"
  }
}
```

But that's not the configuration structure that we need for the pipeline. The `"Input/output options"` is aesthetic only and not structural and the data that we want to validate is as follows:

```json
{ "input": "input.csv" }
```

To get around this, in the Python code I added a hack with a function that "flattens" the schema. It loops through it and moves any properties inside second-level `objects` to the top level and then deletes all `objects`. This works fine, but means that our JSON Schema isn't usable out of the box by others. It also means that it's not possible for people to use more complex schema where they are using objects properly to define actual nested Nextflow parameters.

By using `definitions` to have sub-schema as suggested above, we are no longer fudging the schema to do what we want. We can still define groups of parameters and add keys to each definition for rich form rendering, but we no longer have "fake" `objects` messing up the validation. As such, we can run schema validation with the unchanged schema, which is better for everyone. A sort of equivalent pseudo-schema might look like this:

```yaml
# Input/output options
- input
- single_end

# Reference genome options
- genome
- fasta
```

All parameters which are top-level in the config are now top-level in the schema, so native validation works.

A side effect of making this change is that it will be possible for us to drop our assumption that all parameters are single level deep. So we could start supporting nested config params if we wanted to, using `objects` in the way that they are meant to be used. But I don't think that we should now or for the forseeable future. I think it's good that the standard that we are defining for the Nextflow community of how to describe parameters is as clean and powerful as possible, so that it's future-proofed for others to mirror the same setup as us and add in those more complex features if they need it (and the pipeline validation code etc will work).

Phil

**Comment by mashehu** [comment]
Okay, so with the `allOf` we are kind of symlinking the elements inside the group up to the main level? What I don't completely get is how we can still have the parameters inside `properties`, as shown in the example above. Shouldn't it be:
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "nf-core/gwas pipeline parameters",
  "description": "A pipeline for Genome Wide Association Studies",
  "type": "object",
  "definitions": {
    "inputOutputOptions": {
      "title": "Input/output options",
       "input": {
          "type": "string",
          "fa_icon": "fas fa-dna",
          "description": "Input FastQ files.",
          "help_text": "A glob pattern for input FastQ files. Should include at least one asterisk (*). For paired-end data, should contain curly brackets with two patterns differentiating the paired reads e.g. `*_R{1,2}.fastq.gz`"
        },
        "single_end": {
          "type": "boolean",
          "description": "Specifies that the input is single-end reads.",
          "fa_icon": "fas fa-align-center",
          "default": false,
          "help_text": "By default, the pipeline expects paired-end data. If you have single-end data, specify this parameter on the command line when you launch the pipeline. It is not possible to run a mixture of single-end and paired-end files in one run."
        },
        "outdir": {
          "type": "string",
          "description": "The output directory where the results will be saved.",
          "default": "./results",
          "fa_icon": "fas fa-folder-open"
        },
        "email": {
          "type": "string",
          "description": "Email address for completion summary.",
          "fa_icon": "fas fa-envelope",
          "help_text": "An email address to send a summary email to when the pipeline is completed.",
          "pattern": "^([a-zA-Z0-9_\\-\\.]+)@([a-zA-Z0-9_\\-\\.]+)\\.([a-zA-Z]{2,5})$"
        },
      "required": [
        "input"
      ],
      "fa_icon": "fas fa-terminal",
      "description": "Define where the pipeline should find input data and save output data."
    }
  },
  "allOf": [
    {
      "$ref": "#/definitions/inputOutputOptions"
    }
  ]
}
```

**Comment by ewels** [comment]
Yeah, I don't really know. If you're interested see docs [here](https://json-schema.org/understanding-json-schema/reference/combining.html), [here](https://json-schema.org/understanding-json-schema/structuring.html) and the spec [here](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.2.5).

It looks like both the original example and your example both work with the validation:

✅ https://jsonschema.dev/s/WvHg3 (original)
✅ https://jsonschema.dev/s/YgVET (your version)

Maybe @Relequestual can shed some light on this?
TL;DR - should it be `schema.definitions.groupName.properties.foo` or just  `schema.definitions.groupName.foo` ?

**Comment by ewels** [comment]
Ok I asked about this on the JSON Schema slack and got some help. The misconception that was causing confusion here is about how `allOf` works. It's not gluing the `definitions` together, it's validating each of the `definitions` subschema independently, one after another.

So this:

> _Okay, so with the `allOf` we are kind of symlinking the elements inside the group up to the main level?_

..is not quite right. It's a subtle difference, but it explains how we can have the parameters inside the multiple different `properties` sections but still have them apply to the top-level flat params in the JSON.

Not sure how well I described this. Does it make sense?

**Comment by mashehu** [comment]
okay, so it basically makes every group it's own schema? That would mean we could actually keep them as `objects`, no? Or is the problem with the `object` types not in the validation, but how we use them later on in nextflow?

**Comment by ewels** [comment]
Yes we could keep them as objects, like the top-level schema 👍🏻 We can add that in if you like.

**Comment by ewels** [comment]
Ah you've already got them there in https://github.com/nf-core/tools/pull/688/files - so I think that should be good to go ✅

**Comment by mashehu** [comment]
Good that means that the parsing should actually not change that much 🎉

**Comment by ewels** [comment]
Yes, I guess you're right... Nice!!
